ci: Enable caching for esp32 port builds, split into multiple jobs

projectgus opened the pull request:
Changes to improve performance of esp32 build (previously sitting at 7-8 minutes per run, slowest CI step along with Zephyr.)

1. Cache the ESP-IDF checkout (including all submodules) and the ESP-IDF tools install. Cache is keyed per ESP-IDF version, and the cache action also scopes to each branch (although the default branch cache is available to other branches.) These steps take a little over 1 minute to run on cache miss, but restoring from a cache hit takes approx 15 seconds.
2. Use GitHub Actions build matrix to parameterise which function in ci.sh should be run. This allows splitting up the build steps with minimum copy-paste of boilerplate. With a warm IDF cache the current build jobs each takes around 3 minutes, and will usually run in parallel.
3. Enable [ccache](https://ccache.dev/) for ESP-IDF builds, using hendrikmuhs' [ccache action](https://github.com/hendrikmuhs/ccache-action/) because the cache should be refreshed after each build, the built-in cache action isn't quite enough. A warm ccache, this saves an extra ~40-60 seconds per MicroPython build (i.e. 80-120 seconds per job.) For example see [this job](https://github.com/micropython/micropython/actions/runs/7094430432/job/19309688449?pr=13090), finished in 1m30s.

There is also a commit here to change stm32 build to use the build matrix, for consistency with esp32 rather than because it significantly reduces reduces copy/paste.

projectgus committed:
github/workflows: Cache ESP-IDF checkout and installation.

Cache is keyed on the ESP-IDF version used in CI, so there shouldn't be any
cache invalidation issues when ESP-IDF version changes.

Restoring from cache takes approx 15s, compared to 2-3m to perform these
steps (ESP-IDF tools install, ESP-IDF clone, ESP-IDF submodule clone) the
first time.

Cache size is approx 1.6GB, the git clone is tweaked as much as possible to
keep the size down.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
github/workflows: Use build matrix for esp32 port.

Allows splitting the esp32 job into multiple parts without too much
boilerplate.  The matrix is parameterised using the name of the function to
call in tools/ci.sh, to minimise the dependency on GitHub Actions.

This can get esp32 build times down around 3m if IDF is cached already.

If the cache is cold, the cache preparation step on each job can double up
against each other.  However, restructuring the workflow to not do this
seems either complex or requires copy-pasting the entire cache step.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus committed:
github/workflows: Enable ccache for esp32 build.

CCaches are scoped per-job.

Uses https://github.com/hendrikmuhs/ccache-action to get desired behaviour
(updating the cache on each run).

Signed-off-by: Angus Gratton <angus@redyak.com.au>

dpgeorge commented:
I think it's worth trying out ccache, to get the CI times down even more.

Regarding the second commit, using the build matrix: for the stm32 CI we do it differently and simply have the CI split into two separate jobs, again for speed reasons.  Each job installs the toolchain (doesn't take long) and then performs a set of actions, builds some boards, tests a few other things.  IMO that's pretty simple, simple to understand and simple to extend.  If that were refactored to use a build matrix, I'm not sure how it'd work, it would be rather complex to put each of the CI bits in `tools/ci.sh:ci_stm32_*_build` in a separate matrix stage.

Maybe the esp32 can also work similar to stm32??  Instead of using a build matrix, have a few separate sets of actions to run?

Alternatively, would it be possible for one of the matrix entries to build multiple boards?

My aims here are to improve efficiency of the CI (set up the IDF and build a couple of boards with that set up) and also make it easier to extend in the future (eg run CI commands that are not just building boards).

dpgeorge commented:
We are going to need to add at least one more board to the esp32 CI, see #13131.

Also, we want to keep as much of the CI logic out of the github workflow/action files, and put as much as possible in `tools/ci.sh`.  The reasons are (1) so we can run the CI locally without github actions, by running the functions from `tools/ci.sh`; (2) if we need to migrate to a different CI runner one day then it's relatively easy, the `tools/ci.sh` script can be used as-is (this already happened, when migrating away from Travis).

projectgus committed:
github/workflows: Enable build matrix for stm32 port.

This is for consistency with the previous commit that uses this approach
for esp32.

Signed-off-by: Angus Gratton <angus@redyak.com.au>

projectgus commented:
@dpgeorge Understood. How does the latest version of this PR look, now?

This does still use build matrix, but parameterises it on the CI functions so there's a 1:1 correspondence.

The reason I'm keen to keep the matrix is that it's the least painful way to reuse steps inside GitHub Workflow files without copy/paste or spanning multiple files, and caching requires some additional GitHub Workflow steps.

dpgeorge commented on .github/workflows/ports_esp32.yml:
```diff
@@ -18,11 +23,36 @@ concurrency:
+      matrix:
+        ci_func:  # names are functions in ci.sh
+          - esp32_s2_build
+          - esp32_s3_c3_build
```
I would call them `esp32_build_group1` and `esp32_build_group2` (or set1/set2 or just 1/2), so their contents can change without having to change their name.  The only aim is to make them take about the same time to build.  Then we can add another easily, if/when needed.

projectgus replied:
@dpgeorge Are you sure about this? The `build` functions in `ci.sh` currently have descriptive names (including `ci_stm32_pyb_build` and `ci_stm32_nucleo_build`, I didn't add those names in this PR).

I think that's valuable as an indication of what a particular function will do - some of the build functions are otherwise not particularly self-documenting at first glance.

If we find the name keeps getting longer as we add boards or targets to it, this is also a pretty clear sign that maybe a new job should be added instead.

dpgeorge replied:
> I think that's valuable as an indication of what a particular function will do - some of the build functions are otherwise not particularly self-documenting at first glance.

Yes that's true.

Then I suggest to call them `esp32_build_s3_c3`, ie put "build" first.  Because they are subsets of the build process.

dpgeorge commented on .github/workflows/ports_stm32.yml:
```diff
@@ -18,20 +18,18 @@ concurrency:
+      matrix:
+        ci_func:  # names are functions in ci.sh
+          - stm32_pyb_build
+          - stm32_nucleo_build
```
I would name them similar to the esp32 ones, eg `stm32_build_group1`.

dpgeorge commented on .github/workflows/ports_stm32.yml:
```diff
@@ -18,20 +18,18 @@ concurrency:
-    - name: Build
-      run: source tools/ci.sh && ci_stm32_pyb_build
+    - name: Build ci_${{matrix.ci_func }}
+      run: source tools/ci.sh && ci_${{ matrix.ci_func }}
```
Yeah, this factoring is nice, it eliminates duplication in this YAML file setting up the environment (even though it's pretty simple steps).

projectgus replied:
@dpgeorge Yeah, here it was more a desire to use the same pattern consistently rather than a real win.

projectgus commented:
@dpgeorge PTAL again. Have renamed according to review comments, and also made a change so ci.sh is the authoritative source for `IDF_VER` (no more environment variables in the GitHub Actions YAML files).

dpgeorge commented:
Looks great now, thanks!
